core-container/entry.sh

```diff
@@ -23,145 +23,123 @@ function echoerr() {
    echo "ERROR: $1" >&2
 }
 
-
-function set_provider_adapters() {
-    oldKubeconfig=$KUBECONFIG
-    KUBECONFIG=
-    echo "Waiting for space manager to be ready... this may take a while."
+function wait_kcp_ready() {
+    echo "Waiting for kcp to be ready... this may take a while."
     (
-        until [ "$(kubectl logs $(kubectl get pod --selector=app=kubestellar -o jsonpath='{.items[0].metadata.name}') -c space-manager | grep '***READY***')" != "" ]; do
-            sleep 10
+        until [ "$(kubectl --kubeconfig $kubestellar_kubeconfig logs $(kubectl --kubeconfig $kubestellar_kubeconfig get pod --selector=app=kubestellar -o jsonpath='{.items[0].metadata.name}') -c kcp | grep '***READY***')" != "" ]; do
+           sleep 10
         done
     )
+}
 
-    echo "Delete the kcp provider secret if it already exists, and then create it."
-    if ! kubectl delete secret -n ${NAMESPACE} kcpsec ; then
-        echo "Nothing to delete."
-    fi
-    kubectl get secrets kubestellar -o 'go-template={{index .data "admin.kubeconfig"}}' | base64 --decode > kcpsecret
-    kubectl create secret generic -n ${NAMESPACE} kcpsec --from-file=kubeconfig="kcpsecret"    
-    rm kcpsecret
+function get_kcp_kubeconfig() {
+    wait_kcp_ready
+    while [ $(KUBECONFIG=$kubestellar_kubeconfig kubectl get secret kubestellar | grep -c -e kubestellar) != "1" ]; do 
+        echo "Waiting for kubestellar secret."
+        sleep 1
+    done
+    kcp_kubeconfig_dir="/home/kubestellar/.kcp"
+    kcp_kubeconfig="${kcp_kubeconfig_dir}/admin.kubeconfig"
+    echo "Copying the admin.kubeconfig from kubestellar seret into ${kcp_kubeconfig}..."
+    mkdir -p $kcp_kubeconfig_dir
+    kubectl --kubeconfig $kubestellar_kubeconfig get secrets kubestellar -o jsonpath='{.data.admin\.kubeconfig}' | base64 --decode > $kcp_kubeconfig
+    external_kcp_kubeconfig="${kcp_kubeconfig_dir}/external.kubeconfig"
+    kubectl --kubeconfig $kubestellar_kubeconfig get secrets kubestellar -o jsonpath='{.data.external\.kubeconfig}' | base64 --decode > $external_kcp_kubeconfig
+}
 
-    echo "Delete the kcp provider object if it already exists, and then create it."
-    if ! kubectl delete spaceproviderdesc pkcp ; then
-        echo "Nothing to delete."
-    fi
-    kubectl create -f - <<EOF
+function create_kcp_provider_object() {
+    get_kcp_kubeconfig     # kcp is created in a seperate container 
+    kubectl --kubeconfig $SPACE_MANAGER_KUBECONFIG delete secret -n ${NAMESPACE} kcpsec > /dev/null 2>&1 || true
```

Comment:
Are there scenarios where some concurrent process somewhere in the whole system is concurrently reading this Secret object? If so then this deletion is a problem.

Reply:
Long term we should switch to using the kubestellar secret.  See https://github.com/kubestellar/kubestellar/issues/1336.

Reply:
This function runs in the init container, right? And the space manager runs in a different container. Individual containers can crash and be restarted independently. For example, you have to consider what happens if the space manager container tries to read the Secret while a restarted init container is deleting and re-creating the Secret.